include/zephyr/net/coap_client.h

```diff
@@ -0,0 +1,99 @@
+ * @brief Callback for CoAP request.
+ *
+ * This callback is called for responses to CoAP client requests.
+ * It is used to indicate errors, response codes from server or to deliver payload.
+ * Blockwise transfers cause this callback to be called sequentially with increasing payload offset
+ * and only partial content in buffer pointed by payload parameter.
+ *
+ * @param result_code Result code of the response. Negative if there was a failure in send.
+ *                    @ref coap_response_code for positive.
+ * @param offset Payload offset from the beginning of a blockwise transfer.
+ * @param payload Buffer containing the payload from the response. NULL for empty payload.
+ * @param len Size of the payload.
+ * @param user_data User provided context.
+ * @param last_block Indicates the last block of the response.
+ */
+typedef void (*coap_client_response_cb_t)(uint8_t result_code,
+					  size_t offset, const uint8_t *payload, size_t len,
+					  bool last_block, void *user_data);
+
+/**
+ * @brief Representation of a CoAP client request.
+ */
+struct coap_client_request {
+	enum coap_method method;      /**< Method of the request */
+	bool confirmable;	      /**< CoAP Confirmable/Non-confirmable message */
+	const char *path;	      /**< Path of the requested resource */
+	enum coap_content_format fmt; /**< Content format to be used */
+	uint8_t *payload;	      /**< User allocated buffer for send request */
+	size_t len;		      /**< Length of the payload */
+	coap_client_response_cb_t cb; /**< Callback when response received */
+	void *user_data;	      /**< User provided context */
+};
+
+/** @cond INTERNAL_HIDDEN */
+struct coap_client {
+	int fd;
+	struct sockaddr address;
+	socklen_t socklen;
+	uint8_t send_buf[MAX_COAP_MSG_LEN];
+	uint8_t recv_buf[MAX_COAP_MSG_LEN];
+	uint8_t request_token[COAP_TOKEN_MAX_LEN];
+	int request_tkl;
+	int offset;
+	int retry_count;
+	struct coap_block_context recv_blk_ctx;
+	struct coap_block_context send_blk_ctx;
+	struct coap_pending pending;
+	struct coap_client_request *coap_request;
+	struct coap_packet request;
+	k_tid_t tid;
+	struct k_thread thread;
+	struct k_sem coap_client_recv_sem;
+	atomic_t coap_client_recv_active;
+
+	K_THREAD_STACK_MEMBER(coap_thread_stack, CONFIG_COAP_ASYNC_CLIENT_STACK_SIZE);
+};
+/** @endcond */
+
+/**
+ * @brief Initialize the CoAP client.
+ *
+ * @param[in] client Client instance.
+ * @param[in] info Name for the receiving thread of the client. Setting this NULL will result as
+ *                 default name of "coap_client".
+ *
+ * @return int Zero on success, otherwise a negative error code.
+ */
+int coap_client_init(struct coap_client *client, const char *info);
+
+/**
+ * @brief Send CoAP request
+ *
+ * Operation is handled asynchronously using system's worker queue
+ * or a background thread. Socket must be open and connected to a destination
+ * address. Once the callback is called with result code 0, socket can be closed or
+ * used for another query.
```

Comment:
```suggestion
 * or a background thread. Socket must be open and connected to a destination
 * address. Once the callback is called with result code 0, socket can be closed or
 * used for another query.
```

Is this contradicting the API a bit?
If socket is connected, why do we need the `sockaddr` for?

And when DTLS socket, which is already connected and do not accept any other socket addresses, should we allow `addr` to be NULL?

Reply:
True, latest changes allow now the `addr` to be NULL. Changed the documentation to reflect that.